### PR TITLE
chore(websocket): extract common socket part

### PR DESCRIPTION
**Pull request by pavelfeldman**

### CODE REVIEW

File: src/server/webSocketServer.ts [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+  close() {
+    this._server.close();
+  }
+
+  private _browserMessage(message: ProtocolResponse) {
+    const seqNum = message.id;
+    if (typeof seqNum === 'number') {
+      const request = this._pendingRequests.get(seqNum);
+      if (!request)
+        return;
```
Comment (dgozman) [inline src/server/webSocketServer.ts:80]: assert/debugAssert?

File: src/server/webSocketServer.ts [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+import { helper } from '../helper';
+import { InnerLogger, logError } from '../logger';
+import { ConnectionTransport, ProtocolRequest, ProtocolResponse } from '../transport';
+
+export interface WebSocketServerDelegate {
+  onClientAttached(socket: ws): void;
+  onClientRequest(socket: ws, message: ProtocolRequest): void;
+  onClientDetached(socket: ws): void;
+  onBrowserNotification(message: ProtocolResponse): void;
+  onBrowserResponse(browserId: number, source: ws, message: ProtocolResponse): void;
```
Comment (dgozman) [inline src/server/webSocketServer.ts:28]: Does browserId mean 'json rpc id'?

File: src/server/webkit.ts [hunk 3]
```diff
@@ -88,143 +87,112 @@ export class WebKit extends BrowserTypeBase {
...
+  function removePageProxy(pageProxyId: string): PageProxyData | undefined {
+    const data = pageProxyIds.get(pageProxyId);
+    if (!data)
+      return;
+    pageProxyIds.delete(pageProxyId);
+    return data;
+  }
+
+  const server = new WebSocketServer(transport, logger, port, {
+    onBrowserResponse(seqNum: number, source: ws, message: ProtocolResponse) {
```
Comment (dgozman) [inline src/server/webkit.ts:109]: nit: I'd avoid seqNum here and use "id" to be in protocol terms.

File: src/server/webkit.ts [hunk 4]
```diff
@@ -88,143 +87,112 @@ export class WebKit extends BrowserTypeBase {
...
+    },
 
-    socket.on('message', (message: string) => {
-      const parsedMessage = JSON.parse(Buffer.from(message).toString());
-      const { id, method, params } = parsedMessage;
-      const seqNum = idMixer.generate({ id, socket });
-      transport.send({ ...parsedMessage, id: seqNum });
+    onClientRequest(socket: ws, message: ProtocolRequest) {
+      const { method, params } = message;
+      const id = server.sendMessageToBrowser(message, socket);
```
Comment (dgozman) [inline src/server/webkit.ts:177]: And here 'id' means something else. I guess I don't understand the convention...

File: src/server/webkit.ts [hunk 5]
```diff
@@ -88,143 +87,112 @@ export class WebKit extends BrowserTypeBase {
...
         return;
       }
-      socket.send(JSON.stringify(message));
-      return;
-    }
-    if (method === 'Playwright.pageProxyCreated') {
-      const socket = browserContextIds.get(params.pageProxyInfo.browserContextId);
-      if (!socket || socket.readyState === ws.CLOSING) {
-        // Drop unattributed messages on the floor.
+      if (method === 'Playwright.pageProxyCreated') {
```
Comment (dgozman) [inline src/server/webkit.ts:148]: We have seven 'Playwright.*' events and only handle 3 here :(

File: src/server/firefox.ts [hunk 6]
```diff
@@ -108,39 +107,36 @@ export class Firefox extends BrowserTypeBase {
...
-
-      if (socket.readyState === ws.CLOSING) {
-        if (pendingBrowserContextCreations.has(id)) {
-          transport.send({
-            id: ++SequenceNumberMixer._lastSequenceNumber,
-            method: 'Browser.removeBrowserContext',
-            params: { browserContextId: message.result.browserContextId }
-          });
-        }
+      if (source.readyState === ws.CLOSING) {
```
Comment (dgozman) [inline src/server/firefox.ts:131]: I think we do not handle the case when client detaches after pendingBrowserContextCreations has been sent, but before its response has been received.

File: src/server/firefox.ts [hunk 7]
```diff
@@ -151,88 +147,59 @@ function wrapTransportWithWebSocket(transport: ConnectionTransport, logger: Inne
...
-      return;
-    }
-    if (method === 'Browser.attachedToTarget') {
-      const socket = browserContextIds.get(params.targetInfo.browserContextId);
-      if (!socket || socket.readyState === ws.CLOSING) {
-        // Drop unattributed messages on the floor.
+      if (method === 'Browser.attachedToTarget') {
+        const socket = browserContextIds.get(params.targetInfo.browserContextId);
+        if (!socket || socket.readyState === ws.CLOSING) {
+          // Drop unattributed messages on the floor.
```
Comment (dgozman) [inline src/server/firefox.ts:169]: Should we detach in this case?

File: src/server/chromium.ts [hunk 8]
```diff
@@ -161,111 +157,90 @@ function wrapTransportWithWebSocket(transport: ConnectionTransport, logger: Inne
...
+        return;
       }
-      return;
-    }
 
-    // At this point everything we care about has sessionId.
-    if (!message.sessionId)
-      return;
+      if (message.id !== -1)
+        this.onBrowserNotification(message);
```
Comment (dgozman) [inline src/server/chromium.ts:183]: How can response be a notification?

File: src/server/webSocketServer.ts (hunk 1 above)
Comment (pavelfeldman) [reply src/server/webSocketServer.ts:80]: Can't do - this will knock the server down.

File: src/server/webSocketServer.ts (hunk 2 above)
Comment (pavelfeldman) [reply src/server/webSocketServer.ts:28]: I meant to consistently name it seqNum

File: src/server/webkit.ts (hunk 3 above)
Comment (pavelfeldman) [reply src/server/webkit.ts:109]: Tried that and immediately got confused. I'd like renamed ids to be consistently called seqNums

File: src/server/webkit.ts (hunk 4 above)
Comment (pavelfeldman) [reply src/server/webkit.ts:177]: See, you also got confused because I left one seqNum named id...

File: src/server/chromium.ts (hunk 8 above)
Comment (pavelfeldman) [reply src/server/chromium.ts:183]: ProtocolResponse in transport terms is a JSONRpc response, which is either command response or a notification.

File: src/server/firefox.ts (hunk 6 above)
Comment (pavelfeldman) [reply src/server/firefox.ts:131]: Confused - you are pointing to the line where we `Browser.removeBrowserContext`, what else would you like to happen?

File: src/server/firefox.ts (hunk 7 above)
Comment (pavelfeldman) [reply src/server/firefox.ts:169]: Socket is closing, which means we will run through the close sequence, or what do you mean?

File: src/server/webkit.ts (hunk 5 above)
Comment (pavelfeldman) [reply src/server/webkit.ts:148]: Yes, not sure what you mean by the emoji. Are you saying there is a bug?

File: src/server/webSocketServer.ts (hunk 1 above)
Comment (pavelfeldman) [reply src/server/webSocketServer.ts:80]: Done

File: src/server/webkit.ts (hunk 5 above)
Comment (pavelfeldman) [reply src/server/webkit.ts:148]: Sorted out offline. Blocked by a WebKit patch.

File: src/server/firefox.ts (hunk 6 above)
Comment (pavelfeldman) [reply src/server/firefox.ts:131]: Figured it out

**Review by dgozman — APPROVED** [review]

**Pull request merged**